Short term storage : hour for clusters functions

guilpier-code opened the pull request:
This work is a first refactoring before adding new output variables for ST storage.
It intends to simplify or even remove definition and use of :  
- **hourForEachThermalCluster**
- **hourForEachRenewableCluster**

in order to introduce functions **hourForClusters** to handle clusters of all kinds (thermal, renewable and ST storage).
Eventually, the content of these functions should be moved to **hourForEachArea**.

guilpier-code committed:
[skip ci] st storage output : some cleaning

guilpier-code committed:
st storage output (refactor) : moving code for renewable clusters from top to renewable output variables

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/feature/short-term-storage' into feature/st-storage-output-vars

guilpier-code committed:
st storage output (refactor) : some cleaning

- Class state : some variables should be local instead of public attributes
- Output variables classes : hourForEachThermalCluster removed when empty

guilpier-code committed:
st storage output (refactor) : output variables related to thermal energy manage their clusters output once for all.

Output variables related to thermal energy manage their clusters once output instead of managing them one by one.

guilpier-code committed:
st storage output (refactor) : remove unnecessary hourForEachThermalCluster(...) and hourForEachRenewableCluster(...) member functions

guilpier-code committed:
st storage output (refactor) : renaming hourForEachThermalCluster(...) into hourForThermalClusters(...)

guilpier-code committed:
st storage output (refactor) : removing unused function

flomnes commented on src/solver/variable/state.h:
```diff
@@ -169,24 +162,20 @@ class State
-    double thermalClusterAvailableProduction;
-    //! The operating cost for the current cluster of the current hour (production level*production
+
+    //! Thermal production for thermal clusters for the current hour in the year
+    std::vector<std::vector<double>> thermalClustersProductions;
```
Please make sure to make it precise what the indices should be

flomnes commented on src/solver/variable/state.h:
```diff
@@ -169,24 +162,20 @@ class State
-    double thermalClusterPMinOfAGroup;
-    //! Minimum power of the cluster for the current hour in the year
-    double thermalClusterPMinOfTheCluster;
+    std::vector<std::vector<uint>>  numberOfUnitsONbyCluster;
+    
+    //! Minimum power of all clusters for the current hour in the year
+    std::vector<std::vector<double>> PMinOfClusters;
```
Add comment about indices

flomnes commented on src/solver/variable/state.cpp:
```diff
@@ -46,6 +46,21 @@ State::State(Data::Study& s) :
  averageOptimizationTime1(0.),
  averageOptimizationTime2(0.)
 {
+    thermalClustersProductions.resize(s.areas.size());
+    for (int area_index = 0; area_index < s.areas.size(); area_index++)
+        thermalClustersProductions[area_index].resize(s.areas[area_index]->thermal.clusterCount());
+
+    PMinOfClusters.resize(s.areas.size());
+    for (int area_index = 0; area_index < s.areas.size(); area_index++)
+        PMinOfClusters[area_index].resize(s.areas[area_index]->thermal.clusterCount());
+
+    numberOfUnitsONbyCluster.resize(s.areas.size());
+    for (int area_index = 0; area_index < s.areas.size(); area_index++)
+        numberOfUnitsONbyCluster[area_index].resize(s.areas[area_index]->thermal.clusterCount());
+
+    thermalClustersOperatingCost.resize(s.areas.size());
+    for (int area_index = 0; area_index < s.areas.size(); area_index++)
+        thermalClustersOperatingCost[area_index].resize(s.areas[area_index]->thermal.clusterCount());
```
Introduce a lambda function ? It's 4 times the same code here

flomnes replied:
Note that since C++14, `auto` is allowed for lambda function args.

flomnes commented on src/solver/variable/setofareas.h:
```diff
@@ -150,7 +150,7 @@ class SetsOfAreas
     void hourForEachArea(State& state);
     void hourForEachLink(State& state);
-    void hourForEachThermalCluster(State& state);
+    void hourForThermalClusters(State& state);
```
Why not use `hourForEachArea` ?

guilpier-code committed:
st storage output (refactor) : renaming hourForThermalClusters(...) into hourForClusters(...)

guilpier-code committed:
st storage output (refactor) : fix a regression in results

The pb occurred when renaming functions hourForThermalClusters : apparently, MSVC did not do the job correctly

guilpier-code committed:
st storage output (refactor) : make test python scripts to print the names of failing output variable

guilpier-code committed:
st storage output (refactor) : test python scripts : correction

flomnes committed:
Rename cluster_index -> clusterIndex

flomnes committed:
Apply clang-format

flomnes committed:
Rename cluster_index -> clusterIndex

flomnes committed:
Factorize using lambda function

flomnes committed:
Apply clang-format
